mkosi/__init__.py

```diff
@@ -4190,96 +4210,69 @@ def link_output(args: CommandLineArguments, root: str, artifact: Optional[Binary
             _link_output(args, artifact.name, args.output)
 
 
-def link_output_nspawn_settings(args: CommandLineArguments, path: Optional[str]) -> None:
-    if path is None:
-        return
-
-    assert args.output_nspawn_settings is not None
-
-    with complete_step("Linking nspawn settings file", "Successfully linked " + args.output_nspawn_settings):
-        _link_output(args, path, args.output_nspawn_settings)
-
-
-def link_output_checksum(args: CommandLineArguments, checksum: Optional[str]) -> None:
-    if checksum is None:
-        return
-
-    assert args.output_checksum is not None
-
-    with complete_step("Linking SHA256SUMS file", "Successfully linked " + args.output_checksum):
-        _link_output(args, checksum, args.output_checksum)
-
-
-def link_output_root_hash_file(args: CommandLineArguments, root_hash_file: Optional[str]) -> None:
-    if root_hash_file is None:
-        return
-
-    assert args.output_root_hash_file is not None
-
-    suffix = roothash_suffix(args.usr_only)
-    with complete_step(f"Linking {suffix} file", "Successfully linked " + args.output_root_hash_file):
-        _link_output(args, root_hash_file, args.output_root_hash_file)
-
+def link_output_nspawn_settings(args: CommandLineArguments, path: Optional[SomeIO]) -> None:
+    if path:
+        assert args.output_nspawn_settings
+        with complete_step("Linking nspawn settings file…", f"Successfully linked {args.output_nspawn_settings}"):
+            _link_output(args, path.name, args.output_nspawn_settings)
```

Comment:
I rather prefer checking the bailout condition first and returning instead of adding more indentation, but I guess that's just a matter of preference.

Reply:
True, this repetition has irked me before, too. Maybe some templating?
```python
def _link_output_template(
    args: CommandLineArguments,
    thing: str,
    pretty_name: Optional[str] = None,
): -> None:
    def _link_output_file(args: CommandLineargument, path: Optiona[str]):
        if path is None:
            return None

        output = getattr(args, f"output_{thing}")
        assert output is not None

        with complete_step(f"Linking {pretty_name or thing} file", f"Successfully linked {output}"):
            _link_output(args, path, output)

    return _link_output_file

link_output_checksum = _link_output_template("checksum", "SHA256SUMS")
```
(Haven't tested this, just an idea)

Reply:
Those function are similar… but not identical. So the template would be quite complex. I don't think this is worth it.

Reply:
Yeah, you're right. The ones for roothash and ssh keys are different. Adding hooks back, front and centre to make this work would just be unwieldy.